cursor.go

```diff
@@ -257,6 +257,15 @@ func (c *Cursor) prev() (key []byte, value []byte, flags uint32) {
 			elem.index--
 			break
 		}
+		// If we've hit the beginning, we should stop moving the cursor,
+		// and stay at the first element, so that users can continue to
+		// iterate over the elements in reverse direction by calling `Next`.
+		// We should return nil in such case.
+		// Refer to https://github.com/etcd-io/bbolt/issues/733
+		if len(c.stack) == 1 {
```

Comment:
```
package main

import (
        "fmt"
        "log"

        bolt "go.etcd.io/bbolt"
)

func main() {

        aDb, dbErr := bolt.Open("a.db", 0600, nil)
        if dbErr != nil {
                log.Fatal(dbErr)
        }
        defer aDb.Close()

        aDb.Batch(func(tx *bolt.Tx) error {
                // Create data
                b, _ := tx.CreateBucketIfNotExists([]byte("test"))
                b.Put([]byte("1"), nil)
                b.Put([]byte("2"), nil)
                b.Put([]byte("3"), nil)

                cp := b.Cursor()
                kp, _ := cp.First()
                fmt.Println(string(kp))
                kp, _ = cp.Prev()
                fmt.Println(string(kp))
                kp, _ = cp.Next()
                fmt.Println(string(kp))
                kp, _ = cp.Next()
                fmt.Println(string(kp))
                // End
                return nil
        })
}

```

Based on your change, the result will be like

```
nil <--prev--- 1
 |
 +--next--> 2 --next--> 3
```

The logic looks weird to me actually. If `Next` is like `nil --(next)-->1`, it might make more senses to me.
However, the API looks weird.  Based on the https://pkg.go.dev/go.etcd.io/bbolt#Cursor, before call `Next` or `Prev`, we need to ensure where the position is. When the `Next` or `Prev` is out of range, the position is unknown and the caller should reset the position. I would like to keep it as it is and it's not bug to me.

Reply:
We also need to update the doc to clearly clarify the behaviour of `Next` and `Prev` when the cursor already points the last or first element.

> The logic looks weird to me actually. If Next is like nil --(next)-->1, it might make more senses to me.

The motivation is to make the behaviour of `Prev` to be consistent with `Next`. Currently when the cursor points the last element, `Next` won't move the cursor anymore, and it stays on the last element. But `Prev` will move to nil if the cursor already points to the first element.

> When the Next or Prev is out of range, the position is unknown and the caller should reset the position.

The cursor's position is unknown only before `First` or `Last` is called.

> I would like to keep it as it is and it's not bug to me.

As mentioned in https://github.com/etcd-io/bbolt/issues/733#issuecomment-2068018070, It's a minor issue.

Reply:
> The motivation is to make the behaviour of Prev to be consistent with Next.

OK. The `Prev` removes the `stack` cache for each move, while the `Next` just remains the `stack`.